src/main/java/org/quiltmc/loader/api/config/MetadataType.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.api.config;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * A typed key to be used for setting and getting object metadata.
```

Comment:
As far as I can tell, it's never explained what "object metadata" actually is.

Reply:
I think the idea of "metadata" in general is pretty clear, but I've added some more clarifying javadoc that hopefully helps to paint a better picture.